libraries/chain/account_evaluator.cpp

```diff
@@ -352,19 +356,51 @@ void_result account_update_evaluator::do_apply( const account_update_operation&
 
    bool sa_after = acnt->has_special_authority();
 
+#ifndef ASSET_BALANCE_SORTED
+   flat_set<asset_id_type> assets_after = acnt->get_top_n_control_assets();
+
+   const auto& special_meta = d.get_special_assets_meta_object();
+#endif
+
    if( sa_before && (!sa_after) )
    {
       const auto& sa_idx = d.get_index_type< special_authority_index >().indices().get<by_account>();
       auto sa_it = sa_idx.find( o.account );
       assert( sa_it != sa_idx.end() );
       d.remove( *sa_it );
+#ifndef ASSET_BALANCE_SORTED
+      // From next maintenance interval, need to stop updating this account's active/owner.
+      // Here we check if each asset is added earlier in this maintenance interval,
+      //   if yes, remove it from "added"; otherwise, add it to "removed".
+      d.modify( special_meta, [&assets_before](special_assets_meta_object& p) {
+         auto& added = p.special_assets_added_this_interval;
+         for( const asset_id_type a : assets_before )
+         {
+            if( added.erase( a ) == 0 ) // if not in "added"
+               p.special_assets_removed_this_interval.insert( a );
+         }
+      });
+#endif
    }
    else if( (!sa_before) && sa_after )
    {
       d.create< special_authority_object >( [&]( special_authority_object& sa )
       {
          sa.account = o.account;
       } );
+#ifndef ASSET_BALANCE_SORTED
+      // From next maintenance interval, need to update this account's active/owner, so need to get top n asset holders.
+      // Here we check if each asset is removed earlier in this maintenance interval,
+      //   if yes, remove it from "removed"; otherwise, add it to "added".
+      d.modify( special_meta, [&assets_after](special_assets_meta_object& p) {
+         auto& removed = p.special_assets_removed_this_interval;
+         for( const asset_id_type a : assets_after )
+         {
+            if( removed.erase( a ) == 0 ) // if not in "removed"
+               p.special_assets_added_this_interval.insert( a );
+         }
+      });
+#endif
    }
 
    return void_result();
```

Comment:
IMO the handling of special_assets_meta_object is broken in several respects:
* account_create_evaluator::do_apply can create special authorities, similar code needs to be added there as well
* you don't handle the case if an account switches from one asset to another (i. e. both sa_before and sa_after are true)
* you cannot remove an asset just because an account doesn't use it anymore - it could still be used by a different account

Please add unit tests for all of these.

Reply:
Thanks. My bad. Apparently the code is incomplete. I don't know why I didn't realize it. Perhaps I was thinking that it's updating an asset but not an account.